compiler/rustc_typeck/src/check/wfcheck.rs

```diff
@@ -1518,12 +1537,19 @@ fn check_fn_or_method<'tcx>(
             ty.span,
             Some(WellFormedLoc::Param { function: def_id, param_idx: i.try_into().unwrap() }),
             input_ty.into(),
+            None,
         );
     }
 
     implied_bounds.extend(sig.inputs());
 
-    wfcx.register_wf_obligation(hir_decl.output.span(), None, sig.output().into());
+    // override the env when checking the return type. `~const` bounds can be fulfilled with non-const implementations.
```

Comment:
Shouldn't you also be applying this to arg types too?

Reply:
Also, I guess a deeper question is whether this should be fixed in `rustc_trait_selection::traits::wf` instead, or I guess whether we should be doing wfcheck with a `Constness::Const` param-env at all. (e.g. should we just be doing `param_env.without_const` when constructing the wfcx...?)

Like, is it ever required to register a wf obligation with `Constness::Const` for correctness sake?

Reply:
> whether we should be doing wfcheck with a `Constness::Const` param-env at all?

We should. For an `impl` to be well-formed, all super traits must be fulfilled. In the case of `const` super traits, we need to ensure that the parent trait has a `const` `impl`.

Reply:
I guess I mean for the nominal obligations of the projection types in the signature, for example. Not necessarily when we verify that a trait's supertraits are satisfied. 

Put in a different way, any time we instantiate a WellFormed predicate, is it ever wrong to pass a param env with no constness there instead?

Reply:
That seems desirable. This is related to #100543. I'm going to look into this further.

Reply:
I tend to agree with @compiler-errors' reasoning here (it was my first thought when looking at the code).

Reply:
This has been done, but only from one producer of wf obligations, to fix this issue. The second fix is unfortunate but is the only solution I can find without breaking the tests.